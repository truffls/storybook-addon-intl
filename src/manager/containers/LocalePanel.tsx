import { useChannel, useGlobals } from 'storybook/manager-api';
import React, { Fragment, useState } from 'react';

import { EVENTS, EventsData, GLOBALS_KEY } from '../../constants';
import { parseState } from '../../state';
import { StorybookAddonIntlConfigCommon } from '../../types';
import { getActiveLocale } from '../../utils/locale';
import { LocalePanelContent } from '../components/LocalePanelContent';
import { Warning } from '../components/Warning';
import { Loading } from '../components/Loading';

export type LocalPanelProps = {
    active?: boolean;
};

export function LocalePanel({ active = false }: LocalPanelProps) {
    const [globals, updateGlobals] = useGlobals();
    const rawIntlGlobals = globals[GLOBALS_KEY];
    const intlState = parseState(rawIntlGlobals);

    const [{ config: intlConfig, error: parseError }, setIntlConfigState] =
        useState<{
            config: StorybookAddonIntlConfigCommon | null;
            error: string | null;
        }>({
            config: null,
            error: null
        });

    const locales = intlConfig?.locales ?? [];
    const activeLocale = getActiveLocale(
        locales,
        intlState.activeLocale,
        intlConfig?.defaultLocale
    );

    useChannel({
        [EVENTS.CONFIG_STATUS]: ({
            config,
            error
        }: EventsData['CONFIG_STATUS']) => {
            setIntlConfigState({
                config,
                error
            });
        }
    });

    const changeLocale = (nextActiveLocale: string) => {
        updateGlobals({
            [GLOBALS_KEY]: {
                activeLocale: nextActiveLocale
            }
        });
    };

    return (
        <div hidden={!active}>
            {!intlConfig || !activeLocale || !!parseError ? (
                <Fragment>
                    {!intlConfig || !parseError ? (
                        <Loading>Loading ...</Loading>
                    ) : (
                        <Warning>
                            {parseError === 'missing' ? (
                                <Fragment>
                                    <code>storybook-addon-intl</code> is not
                                    configured.
                                </Fragment>
                            ) : (
                                <Fragment>
                                    <code>storybook-addon-intl</code> is not
                                    configured correctly.
                                    {` (code: ${parseError ?? 'unknown'})`}
                                </Fragment>
                            )}
                        </Warning>
                    )}
                </Fragment>
            ) : (
                <LocalePanelContent
                    locales={locales}
                    activeLocale={activeLocale}
                    onChangeLocale={changeLocale}
                />
            )}
        </div>
    );
}
