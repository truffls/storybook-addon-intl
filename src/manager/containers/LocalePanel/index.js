import React from 'react';
import PropTypes from 'prop-types';
import {
    EVENT_SET_CONFIG_ID,
    EVENT_GET_LOCALE_ID,
    EVENT_SET_LOCALE_ID
} from '../../../shared';

import LocaleButton from '../../components/LocaleButton';

const panelStyle = {
    flexGrow: 0,
    display: 'flex',
    alignSelf: 'flex-start',
    flexWrap: 'wrap'
};

class LocalePanel extends React.Component {
    constructor (props) {
        super(props);

        this.state = {
            locales: [],
            activeLocale: null
        };

        this.setConfig = this.setConfig.bind(this);
        this.getLocale = this.getLocale.bind(this);
        this.handleClickLocaleButton = this.handleClickLocaleButton.bind(this);

        this.props.channel.on(EVENT_SET_CONFIG_ID, this.setConfig);
        this.props.channel.on(EVENT_GET_LOCALE_ID, this.getLocale);
    }

    setConfig (config) {
        const { locales, defaultLocale } = config;

        let activeLocale = this.state.activeLocale;
        // If active locale is not in the list of available locales use default locale as new active
        if (!activeLocale || !~locales.indexOf(activeLocale)) {
            activeLocale = defaultLocale;
        }

        this.setState({
            locales: locales,
            activeLocale: activeLocale
        });
    }

    getLocale () {
        this.props.channel.emit(EVENT_SET_LOCALE_ID, this.state.activeLocale);
    }

    handleClickLocaleButton (locale) {
        this.setState({
            activeLocale: locale
        });

        this.props.channel.emit(EVENT_SET_LOCALE_ID, locale);
    }

    componentWillUnmount () {
        this.props.channel.removeListener(EVENT_SET_CONFIG_ID, this.setConfig);
        this.props.channel.removeListener(EVENT_GET_LOCALE_ID, this.getLocale);
    }

    render () {
        const { active } = this.props;
        const { activeLocale, locales } = this.state;

        if (!active || !locales.length) {
            return null;
        }

        const items = locales.map((locale) => {
            return (
                <LocaleButton
                    key={locale}
                    locale={locale}
                    active={locale === activeLocale}
                    onClick={this.handleClickLocaleButton}
                />
            );
        });

        return (
            <div style={panelStyle}>
                {items}
            </div>
        );
    }
}

LocalePanel.propTypes = {
    active: PropTypes.bool.isRequired,
    channel: PropTypes.shape({
        emit: PropTypes.func.isRequired,
        on: PropTypes.func.isRequired,
        removeListener: PropTypes.func.isRequired
    }).isRequired
};

export default LocalePanel;
