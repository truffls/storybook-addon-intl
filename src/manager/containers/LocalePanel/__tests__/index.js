import React from 'react';
import renderer from 'react-test-renderer';

import LocalePanel from '../';
import {
    EVENT_SET_CONFIG_ID,
    EVENT_GET_LOCALE_ID,
    EVENT_SET_LOCALE_ID
} from '../../../../shared';

describe('LocalePanel', () => {
    test('initialize correctly', () => {
        const channel = {
            emit: () => {},
            on: jest.fn(),
            removeListener: () => {}
        };


        const component = renderer.create(
            <LocalePanel active channel={channel} />
        );
        const instance = component.getInstance();


        expect(instance.state).toEqual({ locales: [], activeLocale: null });

        expect(channel.on.mock.calls).toContainEqual([ EVENT_SET_CONFIG_ID, instance.setConfig ]);
        expect(channel.on.mock.calls).toContainEqual([ EVENT_GET_LOCALE_ID, instance.getLocale ]);
    });

    test('set config correctly', () => {
        const channel = {
            emit: () => {},
            on: () => {},
            removeListener: () => {}
        };


        const component = renderer.create(
            <LocalePanel active channel={channel} />
        );
        const instance = component.getInstance();

        // After initialization
        instance.setConfig({ locales: ['en', 'de'], defaultLocale: 'en' });
        expect(instance.state).toEqual({ locales: ['en', 'de'], activeLocale: 'en' });

        // Initialized
        instance.setState({ locales: ['en', 'fr'], activeLocale: 'en' });

        instance.setConfig({ locales: ['en', 'de'], defaultLocale: 'en' });
        expect(instance.state).toEqual({ locales: ['en', 'de'], activeLocale: 'en' });

        // Initialized - current activeLocale not in locales anymore
        instance.setState({ locales: ['en', 'fr'], activeLocale: 'fr' });

        instance.setConfig({ locales: ['en', 'de'], defaultLocale: 'en' });
        expect(instance.state).toEqual({ locales: ['en', 'de'], activeLocale: 'en' });
    });

    test('emits locale correctly if requested', () => {
        const channel = {
            emit: jest.fn(),
            on: () => {},
            removeListener: () => {}
        };
        const activeLocale = 'en';


        const component = renderer.create(
            <LocalePanel active channel={channel} />
        );
        const instance = component.getInstance();


        instance.setState({ locales: ['en', 'de'], activeLocale: activeLocale });
        instance.getLocale();

        expect(channel.emit.mock.calls[0]).toEqual([ EVENT_SET_LOCALE_ID, activeLocale ]);
    });

    test('handles button click correctly', () => {
        const channel = {
            emit: jest.fn(),
            on: () => {},
            removeListener: () => {}
        };
        const localeToSet = 'de';


        const component = renderer.create(
            <LocalePanel active channel={channel} />
        );
        const instance = component.getInstance();


        instance.setState({ locales: ['en', 'de'], activeLocale: 'en' });

        instance.handleClickLocaleButton(localeToSet);
        expect(instance.state).toEqual({ locales: ['en', 'de'], activeLocale: localeToSet });

        expect(channel.emit.mock.calls[0]).toEqual([ EVENT_SET_LOCALE_ID, localeToSet ]);
    });

    test('unmount correctly', () => {
        const channel = {
            emit: () => {},
            on: () => {},
            removeListener: jest.fn()
        };


        const component = renderer.create(
            <LocalePanel active channel={channel} />
        );
        const instance = component.getInstance();


        component.unmount();
        expect(channel.removeListener.mock.calls).toContainEqual([ EVENT_SET_CONFIG_ID, instance.setConfig ]);
        expect(channel.removeListener.mock.calls).toContainEqual([ EVENT_GET_LOCALE_ID, instance.getLocale ]);
    });

    test('renders correctly when inactive', () => {
        const channel = {
            emit: () => {},
            on: () => {},
            removeListener: jest.fn()
        };

        const component = renderer.create(
            <LocalePanel active={false} channel={channel} />
        );
        const instance = component.getInstance();

        // If active is false we expect null as return value of the render function
        let tree = component.toJSON();
        expect(tree).toBe(null);
    });

    test('renders correctly when active', () => {
        const channel = {
            emit: () => {},
            on: () => {},
            removeListener: jest.fn()
        };


        const component = renderer.create(
            <LocalePanel active channel={channel} />
        );
        const instance = component.getInstance();

        // If not locale is set (which is the inital state) we expect null as return value of the render function
        let tree = component.toJSON();
        expect(tree).toBe(null);


        instance.setState({ locales: ['en', 'de'], activeLocale: 'en' });
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
