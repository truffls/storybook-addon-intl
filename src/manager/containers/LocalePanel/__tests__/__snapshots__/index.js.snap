// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalePanel renders correctly when active 1`] = `
<div
  style={
    Object {
      "alignSelf": "flex-start",
      "display": "flex",
      "flexGrow": 0,
      "flexWrap": "wrap",
    }
  }
>
  <button
    onClick={[Function]}
    style={
      Object {
        "background": "#f7f7f7",
        "border": 0,
        "borderBottom": "1px solid #d3d3d3",
        "borderRight": "1px solid #d3d3d3",
        "fontWeight": "bold",
        "height": "50px",
        "lineHeight": "30px",
        "padding": "5px",
        "textAlign": "center",
        "textTransform": "uppercase",
        "transitionDuration": "100ms",
        "transitionProperty": "background",
        "transitionTimingFunction": "linear",
        "width": "100px",
      }
    }
    type="button"
  >
    en
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "background": "none",
        "border": 0,
        "borderBottom": "1px solid #d3d3d3",
        "borderRight": "1px solid #d3d3d3",
        "height": "50px",
        "lineHeight": "30px",
        "padding": "5px",
        "textAlign": "center",
        "textTransform": "uppercase",
        "transitionDuration": "100ms",
        "transitionProperty": "background",
        "transitionTimingFunction": "linear",
        "width": "100px",
      }
    }
    type="button"
  >
    de
  </button>
</div>
`;
